toolkit/tools/imagegen/installutils/installutils.go

```diff
@@ -72,56 +72,62 @@ func CreateMountPointPartitionMap(partDevPathMap, partIDToFsTypeMap map[string]s
+	// Convert the installMap into a slice of mount points so it can be sorted
+	for mountPoint := range *mountPointMap {
+		// Skip empty mount points
+		if mountPoint == "" {
+			continue
+		}
+		remainingMounts = append(remainingMounts, mountPoint)
+	}
+
+	// We need to make sure we sort the mount points so we don't mount things in the wrong order
+	// e.g.: /dev is mounted before /dev/pts is.
+	if !sortForUnmount {
+		sort.Sort(sort.StringSlice(remainingMounts))
+	} else {
+		// Reverse the sorting so we unmount in the opposite order
+		sort.Sort(sort.Reverse(sort.StringSlice(remainingMounts)))
+	}
+
+	return
+}
+
 // CreateInstallRoot walks through the map of mountpoints and mounts the partitions into installroot
 // - installRoot is the destination path to mount these partitions
 // - mountPointMap is the map of mountpoint to partition device path
 func CreateInstallRoot(installRoot string, mountPointMap, mountPointToMountArgsMap map[string]string) (installMap map[string]string, err error) {
 	installMap = make(map[string]string)
-
-	// Always mount root first
-	err = mountSingleMountPoint(installRoot, rootMountPoint, mountPointMap[rootMountPoint], mountPointToMountArgsMap[rootMountPoint])
-	if err != nil {
-		return
-	}
-	installMap[rootMountPoint] = mountPointMap[rootMountPoint]
-
-	// Mount rest of the mountpoints
-	for mountPoint, device := range mountPointMap {
-		if mountPoint != "" && mountPoint != rootMountPoint {
-			err = mountSingleMountPoint(installRoot, mountPoint, device, mountPointToMountArgsMap[mountPoint])
-			if err != nil {
-				return
-			}
-			installMap[mountPoint] = device
+	for _, mountPoint := range sortMountPoints(&mountPointMap, false) {
+		device := mountPointMap[mountPoint]
+		err = mountSingleMountPoint(installRoot, mountPoint, device, mountPointToMountArgsMap[mountPoint])
+		if err != nil {
+			return
 		}
+		installMap[mountPoint] = device
 	}
 	return
 }
 
 // DestroyInstallRoot unmounts each of the installroot mountpoints in order, ensuring that the root mountpoint is last
 // - installRoot is the path to the root where the mountpoints exist
-// - installMap is the map of mountpoints to partition device paths
-func DestroyInstallRoot(installRoot string, installMap map[string]string) (err error) {
+// - mountPointMap is the map of mountpoints to partition device paths
+func DestroyInstallRoot(installRoot string, mountPointMap map[string]string) (err error) {
 	logger.Log.Trace("Destroying InstallRoot")
-
-	// Convert the installMap into a slice of mount points so it can be sorted
-	var allMountsToUnmount []string
-	for mountPoint := range installMap {
-		// Skip empty mount points
-		if mountPoint == "" {
-			continue
+	// Reverse order for unmounting
+	for _, mountPoint := range sortMountPoints(&mountPointMap, true) {
+		err = diskutils.BlockOnDiskIO(mountPointMap[mountPoint])
+		if err != nil {
+			logger.Log.Panicf("DestroyInstallRoot flush IO Error: %s", err.Error())
```

Comment:
>Panicf [](start = 14, length = 6)

Why panic in this particular case?

Reply:
Hold over from debugging I think... I'll swap to error.

---
In reply to: [552878624](https://github.com/microsoft/CBL-Mariner/pull/481#discussion_r552878624) [](ancestors = 552878624)